# Pursuit of Happiness project - Data Cleaning

In [42]:
import pandas as pd

In [43]:
# FUNCTION DEFINITION: function to create DataFrameMatrix with define row_name and col_name 
def matrix_maker_function(dataframe,row_name,col_name,value_name):
    df = dataframe.filter([row_name,col_name,value_name], axis=1)
    matrix_df = pd.pivot_table(df,index=[row_name],values=[value_name],columns=[col_name], aggfunc='sum')             
    matrix_df = pd.DataFrame(matrix_df.to_records())                                
    matrix_df.columns = [hdr.replace("('", "").replace("', ", "").replace(value_name, "").replace(")", "") \
                     for hdr in matrix_df.columns]
    return matrix_df

In [44]:
def filled_matrix_function(dataframe,row_name,col_name,value_name):
    df = dataframe.filter([row_name,col_name,value_name], axis=1)
    matrix_df = pd.pivot_table(df,index=[row_name],values=[value_name],columns=[col_name], aggfunc='sum')             
    matrix_df = pd.DataFrame(matrix_df.to_records())                                
    matrix_df.columns = [hdr.replace("('", "").replace("', ", "").replace(value_name, "").replace(")", "") \
                     for hdr in matrix_df.columns]
    matrix_df = matrix_df.set_index(row_name).transpose().round(2)
    matrix_df.fillna(method='bfill', inplace=True)
    matrix_df.fillna(method='ffill', inplace=True)
    matrix_df = matrix_df.transpose().reset_index()
    matrix_df = matrix_df[matrix_df[row_name].isin(HappyCountries[row_name])].set_index(row_name)
    matrix_df.drop(matrix_df.columns[[0, 1, 13, 14]], axis = 1, inplace = True)
    return matrix_df

In [45]:
pd.set_option("display.max_rows",None)

# Data evaluation of Happiness Index 

To evaluate the quality of the Happiness Index data, we mold it into a matrix of countries v/s years with matrix_maker_function that I previously created. 
The function uses pivot table to create matrix of value from two factors (or columns). Having this function enable us to quickly create matrixes that we use for various analysis.  

In [46]:
#Read main Happiness Index file - WHR20_DataForTable2.1.xls
happyindex_file = pd.read_excel("../../Data/Resources/WHR20_DataForTable2.1.xls")     
happyindex_file.replace(['Taiwan Province of China'], ['Taiwan'],inplace=True)
country_code = pd.read_csv("../../Data/Resources/countrycode1.csv")

In [47]:
# create matrix for all happiness index
Happy_Index_df = matrix_maker_function(happyindex_file ,'Country name','year','Life Ladder')
print(Happy_Index_df.shape)
Happy_Index_df

(166, 16)


Country name      2005      2006      2007      2008  \
0                  Afghanistan       NaN       NaN       NaN  3.723590   
1                      Albania       NaN       NaN  4.634252       NaN   
2                      Algeria       NaN       NaN       NaN       NaN   
3                       Angola       NaN       NaN       NaN       NaN   
4                    Argentina       NaN  6.312925  6.073158  5.961034   
5                      Armenia       NaN  4.289311  4.881516  4.651972   
6                    Australia  7.340688       NaN  7.285391  7.253757   
7                      Austria       NaN  7.122211       NaN  7.180954   
8                   Azerbaijan       NaN  4.727871  4.568160  4.817189   
9                      Bahrain       NaN       NaN       NaN       NaN   
10                  Bangladesh       NaN  4.318909  4.607322  5.052279   
11                     Belarus       NaN  5.657650  5.616976  5.463332   
12                     Belgium  7.262290       NaN  7.218840  7.116591   
13                      Belize       NaN       NaN  6.450644       NaN   
14                       Benin       NaN  3.329802       NaN  3.667140   
15                      Bhutan       NaN       NaN       NaN       NaN   
16                     Bolivia       NaN  5.373986  5.628419  5.297873   
17      Bosnia and Herzegovina       NaN       NaN  4.899807       NaN   
18                    Botswana       NaN  4.739367       NaN  5.451147   
19                      Brazil  6.636771       NaN  6.320673  6.691425   
20                    Bulgaria       NaN       NaN  3.843798       NaN   
21                Burkina Faso       NaN  3.801491  4.017130  3.846439   
22                     Burundi       NaN       NaN       NaN  3.563228   
23                    Cambodia       NaN  3.568745  4.155971  4.462164   
24                    Cameroon       NaN  3.851072  4.349939  4.291800   
25                      Canada  7.418048       NaN  7.481753  7.485604   
26    Central African Republic       NaN       NaN  4.160130       NaN   
27                        Chad       NaN  3.434801  4.141327  4.632468   
28                       Chile       NaN  6.062852  5.697930  5.789439   
29                       China       NaN  4.560495  4.862862  4.846295   
30                    Colombia       NaN  6.024943  6.138412  6.168395   
31                     Comoros       NaN       NaN       NaN       NaN   
32         Congo (Brazzaville)       NaN       NaN       NaN  3.819792   
33            Congo (Kinshasa)       NaN       NaN       NaN       NaN   
34                  Costa Rica       NaN  7.082465  7.432132  6.850680   
35                     Croatia       NaN       NaN  5.820908       NaN   
36                        Cuba       NaN  5.417869       NaN       NaN   
37                      Cyprus       NaN  6.237958       NaN       NaN   
38              Czech Republic  6.439257       NaN  6.500194       NaN   
39                     Denmark  8.018934       NaN  7.834233  7.970892   
40                    Djibouti       NaN       NaN       NaN  5.009330   
41          Dominican Republic       NaN  5.087968  5.081306  4.842306   
42                     Ecuador       NaN  5.024191  4.995875  5.296513   
43                       Egypt  5.167754       NaN  5.540511  4.631741   
44                 El Salvador       NaN  5.700930  5.295535  5.191494   
45                     Estonia       NaN  5.371055  5.332044  5.451938   
46                    Ethiopia       NaN       NaN       NaN       NaN   
47                     Finland       NaN  7.672449       NaN  7.670627   
48                      France  7.093393  6.582700       NaN  7.008065   
49                       Gabon       NaN       NaN       NaN       NaN   
50                      Gambia       NaN       NaN       NaN       NaN   
51                     Georgia       NaN  3.675108  3.707195  4.156090   
52                     Germany  6.619550       NaN  6.416820  6.521790   
53                    

### Evaluation
We evaluate Happiness Index matrix for its completess. 
The 2005 and 2006 has fairly incomplete data set - thus we eliminate those two years. 
Out of the 166 countries, 38 has more than 5 NaN - we then use the dropna(thresh=10) to eliminate those countries. We ended up with a data frame with the shape of 128 rows and 14 to begin.

In [48]:
# check the completeness of data
Happy_Index_df.isna().sum()

Country name      0
2005            139
2006             77
2007             64
2008             56
2009             52
2010             42
2011             20
2012             24
2013             29
2014             21
2015             23
2016             24
2017             19
2018             24
2019             28
dtype: int64

HappinessIndex data is the most incomplete for 2005 - and 

In [49]:
THappy_Index_df = Happy_Index_df.set_index('Country name').transpose()
THappy_Index_df.isna().sum().head()

Country name
Afghanistan     3
Albania         3
Algeria         7
Angola         11
Argentina       1
dtype: int64

In [50]:
#create the matrix of countries that we are going to study
df1 = Happy_Index_df.drop(Happy_Index_df.columns[1:3], axis=1).round(2)
df1 = df1.dropna(thresh=10).reset_index(drop=True)
print(df1.shape)
df1.to_csv("../../Data/Outputs/HappyIndex_multiyear_mx.csv", index=False)
df1.head(1)

(128, 14)


Country name  2007  2008  2009  2010  2011  2012  2013  2014  2015  2016  \
0  Afghanistan   NaN  3.72   4.4  4.76  3.83  3.78  3.57  3.13  3.98  4.22   

   2017  2018  2019  
0  2.66  2.69  2.38

In [51]:
#check list of rows to be deleted
outcast_df = Happy_Index_df[~Happy_Index_df['Country name'].isin(df1['Country name'])].reset_index(drop=True)
outcast_df.head(1)

Country name  2005  2006  2007  2008  2009      2010      2011      2012  \
0      Algeria   NaN   NaN   NaN   NaN   NaN  5.463567  5.317194  5.604596   

   2013      2014  2015      2016      2017      2018      2019  
0   NaN  6.354898   NaN  5.340854  5.248912  5.043086  4.744627

### Mapping
For vizualization requirements we add 3 letter country codes here and continents later using the mapping function

In [52]:
# create a base data frame of countries that in our country list and the country codes .reset_index(drop=True)
HappyCountries = df1.filter(['Country name'], axis=1)

mapping = dict(country_code[['Country name', 'Code']].values)
HappyCountries['Code'] = country_code['Country name'].map(mapping)
HappyCountries.to_csv("../../Data/Outputs/HappyCountries.csv", index=False)
print(HappyCountries.shape)
HappyCountries.head(1)

(128, 2)


Country name Code
0  Afghanistan  AFG

In [53]:
# drop rows and columns that has too much NaN from original file
new_df = happyindex_file[~happyindex_file['Country name'].isin(outcast_df['Country name'])].round(2)
new_df = new_df[new_df['year'] != (2005,2006)]
new_df = new_df.drop(new_df.columns[9:26], axis=1).reset_index(drop=True)

We create clean matrix (with no Nan) by filling data from succeeding years (if available) with "bfill" and the completed with "ffill" for those that still incomplete. Because the operations can only be performed on columns, the dataframe need to be transpose before the operation and retranspose it afterwards.

In [54]:
HI_df = matrix_maker_function(happyindex_file ,'Country name','year','Life Ladder').set_index("Country name").transpose().round(2)
HI_df.fillna(method='bfill', inplace=True)
HI_df.fillna(method='ffill', inplace=True)
HI_df = HI_df.transpose().reset_index()
HI_df = HI_df[HI_df['Country name'].isin(HappyCountries['Country name'])].set_index("Country name")
HI_df.drop(HI_df.columns[[0, 1, 13, 14]], axis = 1, inplace = True)
HI_df

2007  2008  2009  2010  2011  2012  2013  2014  \
Country name                                                                
Afghanistan                3.72  3.72  4.40  4.76  3.83  3.78  3.57  3.13   
Albania                    4.63  5.49  5.49  5.27  5.87  5.51  4.55  4.81   
Argentina                  6.07  5.96  6.42  6.44  6.78  6.47  6.58  6.67   
Armenia                    4.88  4.65  4.18  4.37  4.26  4.32  4.28  4.45   
Australia                  7.29  7.25  7.45  7.45  7.41  7.20  7.36  7.29   
Austria                    7.18  7.18  7.30  7.30  7.47  7.40  7.50  6.95   
Azerbaijan                 4.57  4.82  4.57  4.22  4.68  4.91  5.48  5.25   
Bahrain                    5.70  5.70  5.70  5.94  4.82  5.03  6.69  6.17   
Bangladesh                 4.61  5.05  5.08  4.86  4.99  4.72  4.66  4.64   
Belarus                    5.62  5.46  5.56  5.53  5.23  5.75  5.88  5.81   
Belgium                    7.22  7.12  6.85  6.85  7.11  6.94  7.10  6.86   
Benin                      3.67  3.67  3.87  3.87  3.87  3.19  3.48  3.35   
Bolivia                    5.63  5.30  6.09  5.78  5.78  6.02  5.77  5.86   
Bosnia and Herzegovina     4.90  4.96  4.96  4.67  4.99  4.77  5.12  5.25   
Botswana                   5.45  5.45  3.55  3.55  3.52  4.84  4.13  4.03   
Brazil                     6.32  6.69  7.00  6.84  7.04  6.66  7.14  6.98   
Bulgaria                   3.84  3.91  3.91  3.91  3.88  4.22  3.99  4.44   
Burkina Faso               4.02  3.85  4.04  4.04  4.79  3.96  3.33  3.48   
Cambodia                   4.16  4.46  4.11  4.14  4.16  3.90  3.67  3.88   
Cameroon                   4.35  4.29  4.74  4.55  4.43  4.24  4.27  4.24   
Canada                     7.48  7.49  7.49  7.65  7.43  7.42  7.59  7.30   
Chad                       4.14  4.63  3.64  3.74  4.39  4.03  3.51  3.46   
Chile                      5.70  5.79  6.49  6.64  6.53  6.60  6.74  6.84   
China                      4.86  4.85  4.45  4.65  5.04  5.09  5.24  5.20   
Colombia                   6.14  6.17  6.27  6.41  6.46  6.37  6.61  6.45   
Congo (Brazzaville)        3.82  3.82  4.51  4.51  4.51  3.92  3.95  4.06   
Costa Rica                 7.43  6.85  7.61  7.27  7.23  7.27  7.16  7.25   
Croatia                    5.82  5.43  5.43  5.60  5.39  6.03  5.89  5.38   
Cyprus                     6.83  6.83  6.83  6.39  6.69  6.18  5.44  5.63   
Czech Republic             6.50  6.25  6.25  6.25  6.33  6.33  6.70  6.48   
Denmark                    7.83  7.97  7.68  7.77  7.79  7.52  7.59  7.51   
Dominican Republic         5.08  4.84  5.43  4.74  5.40  4.75  5.02  5.39   
Ecuador                    5.00  5.30  6.02  5.84  5.80  5.96  6.02  5.95   
Egypt                      5.54  4.63  5.07  4.67  4.17  4.20  3.56  4.89   
El Salvador                5.30  5.19  6.84  6.74  4.74  5.93  6.33  5.86   
Estonia                    5.33  5.45  5.14  5.49  5.49  5.36  5.37  5.56   
Finland                    7.67  7.67  7.39  7.39  7.35  7.42  7.44  7.38   
France                     7.01  7.01  6.28  6.80  6.96  6.65  6.67  6.47   
Gabon                      4.26  4.26  4.26  4.26  4.26  3.97  3.80  3.92   
Georgia                    3.71  4.16  3.80  4.10  4.20  4.25  4.35  4.29   
Germany                    6.42  6.52  6.64  6.72  6.62  6.70  6.97  6.98   
Ghana                      5.22  4.97  4.20  4.61  5.61  5.06  4.97  3.86   
Greece                     6.65  6.04  6.04  5.84  5.37  5.10  4.72  4.76   
Guatemala                  6.33  6.41  6.45  6.29  5.74  5.86  5.98  6.54   
Guinea                     4.04  4.04  4.04  4.04  4.04  3.65  3.90  3.41   
Haiti                      3.85  3.85  3.77  3.77  4.84  4.41  4.62  3.89   
Honduras                   5.10  5.42  6.03  5.87  4.96  4.60  4.71  5.06   
Hong Kong S.A.R. of China  5.14  5.14  5.40  5.64  5.47  5.48  5.46  5.46   
Hungary                    4.95  4.89  4.89  4.73  4.92  4.68  4.91  5.18   
India                      5.03  5.15  4.52  4.99  4.63  4.72  4.43  4.42   
Indonesia              

In [55]:
#create a list of indicators
indicators = new_df.keys()[2:10]
names = ['Happy_Index_mx','Log_GDP_mx','Soc_support_mx','Life_exp_mx','Freedom_mx','Generosity_mx','Corruption_mx']

print(indicators)
print(names)

Index(['Life Ladder', 'Log GDP per capita', 'Social support',
       'Healthy life expectancy at birth', 'Freedom to make life choices',
       'Generosity', 'Perceptions of corruption'],
      dtype='object')
['Happy_Index_mx', 'Log_GDP_mx', 'Soc_support_mx', 'Life_exp_mx', 'Freedom_mx', 'Generosity_mx', 'Corruption_mx']


In [56]:
#create a list of indicators
indicators = new_df.keys()[2:10]

#for all indicators create matrixes of country v/s year
for i in range (0,len(indicators)):
    df = filled_matrix_function(new_df,'Country name','year',indicators[i])
    df.to_csv(f"../../Data/Outputs/New_HI{i}_mx.csv", index=False)

# 2017 Basic Data Frame

#### Our group decided to focus the evaluation on the 2017 data. 
To conduct the analysis and create visualization, we need a compilation dataframe of the Happiness Index data and other evaluation factors.

We started with creating basic dataframe of the 128 countries, the Country Name, Code, Happiness Index and its 6 factors, as its columns. We use the shape of this dataframe as the compliance measurement as we build our compilation dataframe.

In [57]:
H1 = HappyCountries.copy()
df_2017 = new_df[new_df["year"]==2017]
df_2017 = pd.merge(H1, df_2017, how='left', on=('Country name')).reset_index(drop=True)
df_2017.rename(columns={"Life Ladder":"Happiness Index","Log GDP per capita":"Log GDP","Healthy life expectancy at birth":"Healthy life exp",
             "Freedom to make life choices":"Freedom","Perceptions of corruption":"Pct corruption"},inplace=True)
df_2017.drop(df_2017.columns[[2]], axis = 1, inplace = True)
print(df_2017.shape) 
df_2017.head(1)         

(128, 9)


Country name Code  Happiness Index  Log GDP  Social support  \
0  Afghanistan  AFG             2.66     7.47            0.49   

   Healthy life exp  Freedom  Generosity  Pct corruption  
0              52.8     0.43       -0.11            0.95

# GHDx Data

In [58]:
#Read main file - IHME-GBD_2017_DATA-ba7d53b0-1 (without age group)
GHDx_file = pd.read_csv("../../Data/Resources/IHME-GBD_2017_DATA-ba7d53b0-1.csv")            
GHDx_file.replace(['Taiwan (Province of China)'], ['Taiwan'],inplace=True)
GHDx_map = pd.read_csv("../../Data/Resources/GHDx_map.csv")

Choose appropriate data and put all together in a data frame

In [59]:
# DEATH BY SUICIDE RATE/100,000 pop MATRIX
# measure_id = 1 Death, sex_id = 3 All, CauseNo = 718 (selfharm/suicide), MetricId = 3 Rate , year = 2017
suicide_df = GHDx_file[(GHDx_file["measure_id"]==1) & (GHDx_file["sex_id"]==3) & (GHDx_file["cause_id"] == 718) & 
                  (GHDx_file["metric_id"]==3) & (GHDx_file["year"]==2017) ].round(2).reset_index()

# DEPRESSION AND ANXIETY (by Incidents) RATE/100,000 pop MATRIX 
# measure_id = 6 Incidence, sex_id = 3 All, CauseNo = 571 Anxiety disorders, year = 2017
anx_df = GHDx_file[(GHDx_file["measure_id"]==6) & (GHDx_file["sex_id"]==3) & (GHDx_file["cause_id"] == 571) & 
                  (GHDx_file["metric_id"]==3) & (GHDx_file["year"]==2017) ].round(2).reset_index()

# measure_id = 6 Incidence, sex_id = 3 All, CauseNo = 567 Depressive disorders, year = 2017
dep_df = GHDx_file[(GHDx_file["measure_id"]==6) & (GHDx_file["sex_id"]==3) & (GHDx_file["cause_id"] == 567) & 
                  (GHDx_file["metric_id"]==3) & (GHDx_file["year"]==2017) ].round(2).reset_index()

# SUBSTANCE ABUSE (by Incidents) RATE/100,000 pop MATRIX 
# measure_id = 6 Incidence, sex_id = 3 All, CauseNo = 973 Substance Abuse, year = 2017
subst_df = GHDx_file[(GHDx_file["measure_id"]==6) & (GHDx_file["sex_id"]==3) & (GHDx_file["cause_id"] == 973) & 
                  (GHDx_file["metric_id"]==3) & (GHDx_file["year"]==2017) ].round(2).reset_index()

GHDx_df = suicide_df.filter(['location_name','val'], axis=1)
GHDx_df['Depression & Anxiety'] = anx_df["val"] + dep_df["val"] 
GHDx_df['Substance Abuse'] = subst_df["val"] 
GHDx_df = GHDx_df.sort_values(by="location_name").rename(columns={"val":"Suicide"}).reset_index(drop=True)

#GHDx_df.to_csv("../Outputs/GHDx_df.csv", index=False)
print(GHDx_df.shape)
GHDx_df.head(1)

(195, 4)


location_name  Suicide  Depression & Anxiety  Substance Abuse
0   Afghanistan     5.25               3870.85           687.47

In [60]:
mapping1 = dict(GHDx_map[['location_name', 'Code']].values)
GHDx_df.insert(1, 'Code', GHDx_map['location_name'].map(mapping1))

#### Columns are added to the df_2017 dataframe with "left-merge" 
This method is choosen to retain the basic dataframe shape and to allow the not existence values be entered as "NaN".  

Here we add the GHDx data into the main df_2017 data frame. Afterwards we double check on the countries that has the NaN value if they are not being included just because of spelling differences.

In [61]:
df_2017 = pd.merge(df_2017 , GHDx_df, how='left', on='Code')
df_2017.drop(df_2017.columns[9], axis=1, inplace=True)
print(df_2017.shape)
df_2017.head()

(128, 12)


Country name Code  Happiness Index  Log GDP  Social support  \
0  Afghanistan  AFG             2.66     7.47            0.49   
1      Albania  ALB             4.64     9.38            0.64   
2    Argentina  ARG             6.04     9.85            0.91   
3      Armenia  ARM             4.29     9.08            0.70   
4    Australia  AUS             7.26    10.71            0.95   

   Healthy life exp  Freedom  Generosity  Pct corruption  Suicide  \
0              52.8     0.43       -0.11            0.95     5.25   
1              68.4     0.75       -0.03            0.88     5.86   
2              68.6     0.83       -0.18            0.84    11.35   
3              66.6     0.61       -0.13            0.86     9.97   
4              73.3     0.91        0.31            0.41    12.76   

   Depression & Anxiety  Substance Abuse  
0               3870.85           687.47  
1               3179.85           642.91  
2               3666.08          1089.71  
3               4525.35           736.38  
4               3291.17           455.44

In [62]:
#check list of rows to be deleted
outcast_df1 = GHDx_df[~GHDx_df['Code'].isin(df_2017['Code'])].reset_index(drop=True)
outcast_df1.head(1)

location_name Code  Suicide  Depression & Anxiety  Substance Abuse
0       Algeria  DZA      4.2               4155.35            560.9

# TheGlobalEconomy Data

Most of our variable came from TheGlobalEconomy Data, which were downloaded and came as two separated csv files.
The lengthy named columns are renamed, then we picked the year and combined them into one data frame, which later be merged with the df_2017 main dataframe.
This data has the same country code, which make the data picking fairly easy.

In [63]:
Global1_file = pd.read_csv("../../Data/Resources/TheGlobalEconomy1.csv")  
Global2_file = pd.read_csv("../../Data/Resources/TheGlobalEconomy2.csv") 

In [64]:
Global1_file.rename(columns={"Gini income inequality index":"Gini index","Poverty percent of population":"Pct Poverty",
                            "Public spending on education percent of public spending":"Pct spent ed",
                             "International tourism revenue":"Rev tourism"},inplace=True)
Global2_file.rename(columns={"Fragile state index 0 (low) - 120 (high)":"Fragile index",
                             "Security threats index 0 (low) - 10 (high)":"Security index",
                             "Human flight and brain drain index 0 (low) - 10 (high)":"brain drain"},inplace=True)

In [65]:
Global1_df = Global1_file[Global1_file["Year"]==2017]
Global1_df = Global1_df.filter(['Code','Gini index','Pct Poverty','Pct spent ed','Literacy rate','Tourist arrivals','Rev tourism'], axis=1).reset_index(drop=True)  
Global_df = pd.merge(HappyCountries, Global1_df, how='left', on=('Code')).reset_index(drop=True)

Global2_df = Global2_file[Global2_file["Year"]==2017]
Global2_df = Global2_df.filter(['Code','Fragile State index','Security index','Brain drain','Unemployment rate'], axis=1).reset_index(drop=True)  
Global_df = pd.merge(Global_df, Global2_df, how='left', on=('Code')).reset_index(drop=True)

print(Global_df.shape)
Global_df.head(1)

(128, 10)


Country name Code  Gini index  Pct Poverty  Pct spent ed  Literacy rate  \
0  Afghanistan  AFG         NaN          NaN         15.66            NaN   

   Tourist arrivals  Rev tourism  Security index  Unemployment rate  
0               NaN         16.0            10.0              11.18

In [66]:
df_2017 = pd.merge(df_2017, Global_df, how='left', on=('Country name','Code')).reset_index(drop=True)
print(df_2017.shape)
df_2017.head(1)

(128, 20)


Country name Code  Happiness Index  Log GDP  Social support  \
0  Afghanistan  AFG             2.66     7.47            0.49   

   Healthy life exp  Freedom  Generosity  Pct corruption  Suicide  \
0              52.8     0.43       -0.11            0.95     5.25   

   Depression & Anxiety  Substance Abuse  Gini index  Pct Poverty  \
0               3870.85           687.47         NaN          NaN   

   Pct spent ed  Literacy rate  Tourist arrivals  Rev tourism  Security index  \
0         15.66            NaN               NaN         16.0            10.0   

   Unemployment rate  
0              11.18

# Other Files

We added other datas that is being evaluated such as GDP, number of Physician perpop, Average Age, and sleep per (min) 

In [67]:
# GDP PPP PerPopulation
gdp_df = pd.read_csv("../../Data/Resources/gdp_ppp_pop.csv", encoding='cp1252')
gdp_df = gdp_df.filter(["Country Name","Country Code","2017"], axis=1).round(2)

gdp_df = pd.merge(HappyCountries, gdp_df, how='left', left_on='Code', right_on='Country Code',
         left_index=False, right_index=False, sort=True)

df_2017.insert(2,('GDP'),gdp_df['2017'])

print(df_2017.shape)
df_2017.head(1)

(128, 21)


Country name Code      GDP  Happiness Index  Log GDP  Social support  \
0  Afghanistan  AFG  2202.57             2.66     7.47            0.49   

   Healthy life exp  Freedom  Generosity  Pct corruption  ...  \
0              52.8     0.43       -0.11            0.95  ...   

   Depression & Anxiety  Substance Abuse  Gini index  Pct Poverty  \
0               3870.85           687.47         NaN          NaN   

   Pct spent ed  Literacy rate  Tourist arrivals  Rev tourism  Security index  \
0         15.66            NaN               NaN         16.0            10.0   

   Unemployment rate  
0              11.18  

[1 rows x 21 columns]

In [68]:
# Physician perpop
#url'https://data.worldbank.org/indicator/SH.MED.PHYS.ZS'
df_hcd=pd.read_csv('../../Data/Resources/doctors_population_WB1-2019.csv')
df_hcd = df_hcd.filter(["Country Name","Country Code","2017"], axis=1).round(2)
df_hcd = pd.merge(HappyCountries, df_hcd, how='left', left_on='Code', right_on='Country Code')

df_2017.insert(13,('Physician perpop'),df_hcd['2017'])

print(df_2017.shape)
df_2017.head(1)

(128, 22)


Country name Code      GDP  Happiness Index  Log GDP  Social support  \
0  Afghanistan  AFG  2202.57             2.66     7.47            0.49   

   Healthy life exp  Freedom  Generosity  Pct corruption  ...  \
0              52.8     0.43       -0.11            0.95  ...   

   Substance Abuse  Physician perpop  Gini index  Pct Poverty  Pct spent ed  \
0           687.47               NaN         NaN          NaN         15.66   

   Literacy rate  Tourist arrivals  Rev tourism  Security index  \
0            NaN               NaN         16.0            10.0   

   Unemployment rate  
0              11.18  

[1 rows x 22 columns]

In [69]:
# Average Age
Age_file = pd.read_csv("../../Data/Resources/Avg.age2017-final.csv") 

del Age_file['Column1']
Age_file.rename(columns={ Age_file.columns[0]: "Country Name", Age_file.columns[1]:"Average age" }, inplace = True)
Age_file['Country Name'] = Age_file['Country Name'].str.strip()

Age_df = pd.merge(HappyCountries, Age_file, how='left', left_on='Country name',  #fit data to standard dataframe
         right_on='Country Name',left_index=False, right_index=False, sort=True)

df_2017.insert(14,('Average Age'),Age_df['Average age'])                        #add to df_2017 dataframe
print(df_2017.shape)
df_2017.head(1)

(128, 23)


Country name Code      GDP  Happiness Index  Log GDP  Social support  \
0  Afghanistan  AFG  2202.57             2.66     7.47            0.49   

   Healthy life exp  Freedom  Generosity  Pct corruption  ...  \
0              52.8     0.43       -0.11            0.95  ...   

   Physician perpop  Average Age  Gini index  Pct Poverty  Pct spent ed  \
0               NaN         18.8         NaN          NaN         15.66   

   Literacy rate  Tourist arrivals  Rev tourism  Security index  \
0            NaN               NaN         16.0            10.0   

   Unemployment rate  
0              11.18  

[1 rows x 23 columns]

In [70]:
# Sleep
Sleep_file = pd.read_csv("../../Data/Resources/Avg_sleep_minutes.csv") 

sleep_df = pd.merge(HappyCountries, Sleep_file, how='left', left_on='Country name',  #fit data to standard dataframe
         right_on='Country',left_index=False, right_index=False, sort=True)

df_2017.insert(15,('Sleep (min)'),sleep_df['Minutes'])                        #add to df_2017 dataframe
print(df_2017.shape)
df_2017.head(1)

(128, 24)


Country name Code      GDP  Happiness Index  Log GDP  Social support  \
0  Afghanistan  AFG  2202.57             2.66     7.47            0.49   

   Healthy life exp  Freedom  Generosity  Pct corruption  ...  Average Age  \
0              52.8     0.43       -0.11            0.95  ...         18.8   

   Sleep (min)  Gini index  Pct Poverty  Pct spent ed  Literacy rate  \
0          NaN         NaN          NaN         15.66            NaN   

   Tourist arrivals  Rev tourism  Security index  Unemployment rate  
0               NaN         16.0            10.0              11.18  

[1 rows x 24 columns]

In [71]:
# Create output file of df_2017
df_2017.to_csv("../../Data/Outputs/df_2017.csv", index=True)

# Normalization

Because the factors that are use to evaluate have different values on totally different scale, to be able to compare them, we have to normalize the values into comparable values. 

We determine the max and minimum of each variable, calculate their ranges, then project all values into a the scale between 0 to 1.

In [72]:
min_value = df_2017.min()
max_value = df_2017.max()

norml_df = pd.DataFrame({'Min Value':min_value,'Max Value':max_value})
norml_df.drop(norml_df.index[:3], inplace=True)
norml_df['range'] = norml_df['Max Value'] - norml_df['Min Value']

norml_df.to_csv("../../Data/Outputs/Normalization value 2017.csv", index=True)
norml_df

Min Value Max Value    range
Happiness Index           2.66      7.79     5.13
Log GDP                   6.83     11.44     4.61
Social support            0.44      0.96     0.52
Healthy life exp          47.7      76.5     28.8
Freedom                   0.43      0.99     0.56
Generosity               -0.29      0.48     0.77
Pct corruption            0.16      0.95     0.79
Suicide                   2.86     36.27    33.41
Depression & Anxiety   1843.64   7022.08  5178.44
Substance Abuse         296.64   2282.46  1985.82
Physician perpop          0.04      6.13     6.09
Average Age               15.4      47.3     31.9
Sleep (min)                377       458       81
Gini index                25.4      53.3     27.9
Pct Poverty                2.4        70     67.6
Pct spent ed              6.95      24.4    17.45
Literacy rate             48.3     99.79    51.49
Tourist arrivals            11     86758    86747
Rev tourism                 16    251544   251528
Security index               1        10        9
Unemployment rate         0.49     27.07    26.58

In [73]:
col_2017 = df_2017.keys()
print(col_2017)

Index(['Country name', 'Code', 'GDP', 'Happiness Index', 'Log GDP',
       'Social support', 'Healthy life exp', 'Freedom', 'Generosity',
       'Pct corruption', 'Suicide', 'Depression & Anxiety', 'Substance Abuse',
       'Physician perpop', 'Average Age', 'Sleep (min)', 'Gini index',
       'Pct Poverty', 'Pct spent ed', 'Literacy rate', 'Tourist arrivals',
       'Rev tourism', 'Security index', 'Unemployment rate'],
      dtype='object')


In [74]:
#NORMALIZATION TABLE 2017
norm_2017_df = country_code.copy()   
for i in range(3,len(col_2017)):
    norm_2017_df[col_2017 [i]]= (df_2017[col_2017 [i]] - norml_df.loc[col_2017 [i],'Min Value']) / norml_df.loc[col_2017 [i],'range']

norm_2017_df['Blue Index'] = norm_2017_df[['Suicide','Depression & Anxiety','Substance Abuse']].mean(axis=1)
norm_2017_df.to_csv("../../Data/Outputs/Normalized Table 2017_df.csv", index=False)
norm_2017_df.round(2)

Country name Code Continent  Happiness Index  Log GDP  \
0                  Afghanistan  AFG      Asia             0.00     0.14   
1                      Albania  ALB    Europe             0.39     0.55   
2                    Argentina  ARG  Americas             0.66     0.66   
3                      Armenia  ARM      Asia             0.32     0.49   
4                    Australia  AUS   Oceania             0.90     0.84   
5                      Austria  AUT    Europe             0.90     0.85   
6                   Azerbaijan  AZE      Asia             0.49     0.62   
7                      Bahrain  BHR      Asia             0.70     0.84   
8                   Bangladesh  BGD      Asia             0.32     0.30   
9                      Belarus  BLR    Europe             0.56     0.63   
10                     Belgium  BEL    Europe             0.83     0.83   
11                       Benin  BEN    Africa             0.43     0.17   
12                     Bolivia  BOL  Americas             0.58     0.43   
13      Bosnia and Herzegovina  BIH    Europe             0.47     0.56   
14                    Botswana  BWA    Africa             0.16     0.62   
15                      Brazil  BRA  Americas             0.72     0.59   
16                    Bulgaria  BGR    Europe             0.48     0.65   
17                Burkina Faso  BFA    Africa             0.39     0.13   
18                    Cambodia  KHM      Asia             0.38     0.30   
19                    Cameroon  CMR    Africa             0.47     0.28   
20                      Canada  CAN  Americas             0.93     0.84   
21                        Chad  TCD    Africa             0.37     0.14   
22                       Chile  CHL  Americas             0.71     0.69   
23                       China  CHN      Asia             0.48     0.61   
24                    Colombia  COL  Americas             0.68     0.58   
25         Congo (Brazzaville)  COG    Africa             0.43     0.37   
26                  Costa Rica  CRI  Americas             0.89     0.61   
27                     Croatia  HRV    Europe             0.52     0.70   
28                      Cyprus  CYP      Asia             0.66     0.78   
29              Czech Republic  CZE    Europe             0.81     0.77   
30                     Denmark  DNK    Europe             0.96     0.85   
31          Dominican Republic  DOM  Americas             0.58     0.60   
32                     Ecuador  ECU  Americas             0.62     0.52   
33                       Egypt  EGY    Africa             0.25     0.53   
34                 El Salvador  SLV  Americas             0.72     0.45   
35                     Estonia  EST    Europe             0.64     0.75   
36                     Finland  FIN    Europe             1.00     0.82   
37                      France  FRA    Europe             0.78     0.81   
38                       Gabon  GAB    Africa             0.41     0.62   
39                     Georgia  GEO      Asia             0.35     0.51   
40                     Germany  DEU    Europe             0.86     0.84   
41                       Ghana  GHA    Africa             0.55     0.32   
42                      Greece  GRC    Europe             0.49     0.71   
43                   Guatemala  GTM  Americas             0.72     0.45   
44                      Guinea  GIN    Africa             0.43     0.19   
45                       Haiti  HTI  Americas             0.23     0.13   
46                    Honduras  HND  Americas             0.65     0.34   
47   Hong Kong S.A.R. of China  HKG      Asia             0.53     0.89   
48                     Hungary  HUN    Europe             0.66     0.73   
49                       India  IND      Asia             0.27     0.42   
50                   Indonesia  IDN      Asia             0.48     0.54   
51                        Iran  IRN      Asia             0.40     0.66   
52                        Iraq  IRQ      As

# Adding continents

For visualization purposes we added continents to our normalized dataframe

In [75]:
norm_2017_v2_df = norm_2017_df.copy()
mapping1 = dict(country_code[['Code','Continent']].values)
norm_2017_v2_df['Continent'] = country_code['Code'].map(mapping1)

norm_2017_v2_df.to_csv("../../Data/Outputs/Normalized Table 2017_df-v2.csv", index=False)

print(norm_2017_v2_df.shape)
norm_2017_v2_df.head(1)

(128, 25)


Country name Code Continent  Happiness Index   Log GDP  Social support  \
0  Afghanistan  AFG      Asia              0.0  0.138829        0.096154   

   Healthy life exp  Freedom  Generosity  Pct corruption  ...  Sleep (min)  \
0          0.177083      0.0    0.233766             1.0  ...          NaN   

   Gini index  Pct Poverty  Pct spent ed  Literacy rate  Tourist arrivals  \
0         NaN          NaN       0.49914            NaN               NaN   

   Rev tourism  Security index  Unemployment rate  Blue Index  
0          0.0             1.0           0.402182    0.219939  

[1 rows x 25 columns]